Maintenance Windows RFC

mcalmer opened the pull request:
How to implement Maintenance Windows in SUSE Manager.

References:
- docs about ITSM tools - https://docs.servicenow.com/search?q=Create+a+maintenance+plan
- https://checkmk.com/cms_basics_downtimes.html

hustodemon commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,99 @@
+## Maintenance Schedule
+
+A Maintenance Schedule is a set of Maintenance Windows. A Maintenance Schedule has a name/label.
+The data should be stored as ICal data (https://tools.ietf.org/html/rfc5545).
```
I wonder if we want to store the iCal data as a varchar in the db. If we do it this way, we need to be prepared that some queries, we might need, can be costly (in general, it would mean selecting everything to memory and do the filtering on backend).

mcalmer replied:
Or we convert ical data as single events into a DB structure. This requires of cause update capability. On the other hand it would make it easier to get these data also in python and other languages when needed.

parlt91 replied:
If we convert the iCal into single events how would we do the updating? I guess we don't want to override the entire table every time. (Could do a check what changed on frontend. What about xmlrpc?)

mbrookhuis replied:
It should be possible to create the schedule via the API

mcalmer replied:
Added some notes about XMLRPC API usage to the RFC.

hustodemon commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,99 @@
+It has a start date-time and an end date-time.
+It can be defined as recurring interval.
+
+When a Maintenance Window start, it turn the maintenance mode "on".
```
We also need to clarify "how" does the switch happen. We need some mechanism, that is "triggered" when the maint. window starts/ends. The first idea that comes to my mind is that when we create the maintenance window, we need to create 2 taskomatic schedules (1 for turning on the maint. mode for the systems, 1 for turning it off). The one problem I could foresee now is that taskomatic only support cron expressions and I fear iCal allows defining more elaborate frequencies (last day of month).

mcalmer replied:
Do we really need a "value" or could it be a function `isInMaintenanceMode()` which lookup the calendar to decide if it is or not?

parlt91 replied:
Agreed we could perform this check every time when scheduling an action and would not need this `maintenance mode` flag. The org admin could get a warning in this case (could decide to perform the action anyway), other users would get an error stating that the system isn't in `maintenance mode` at that time.

joesusecom replied:
From a usage point of view, there are two different interactions I can imagine:

- admin defines a maintenance schedule and expects the system to be in maintenance mode at the given times
- admin (or an API call triggered by a third party tool/scheduler) puts the system into planned or emergency maintenance mode _now_ or switches it off again.

Of course you can implement both in different ways. For example, if there's only a method to check whether the system has a maintenance mode every time an action is triggered, you'll need to create a (temporary) schedule when the system is set to maintenance manually.

But it may be easier to do it the other way around: Manual maintenance mode changes the flag immediately, and the schedules are checked at certain intervals to set/unset maintenance mode as needed.

You may want to look into the potential performance implications of both, for a typical scenario (i.e. is it going to create more load to set/unset the flags for thousands of machines when a window is hit or to calculate the status on the fly based on the schedule). There must be prior art. Calendaring tools are doing this (busy status) all the time. 😄 

Additional considerations:

- Would we want to bring back a system from maintenance mode only at the end of the window or should there be away of switching it off once all maintenance tasks have been finished (which could be much earlier than the end of the window)? I'd think that most admins would _love_ to have the systems come back sooner (which could also be used in uptime statistics to prove that SLAs are met/improving). But the corner case is when an admin figures out later (but still within the maintenance window originally scheduled) that there's more to do.
- What happens if a system is in manual maintenance mode and the scheduled maintenance window kicks in? I guess an admin wouldn't want a system to come back from maintenance automatically at the end of the window if the reason for the manual override hasn't been resolved yet.
- What if a long-running task _exceeds_ the scheduled window and can't be rolled back quickly? In other words, unless the task gets finished, the system will be in an unusable state. Yes, this should never happen. But if it _does_, I'd think the admin would rather see the system flagged as in maintenance until things are finished than get the system back to production while it's still broken.

In the end, this could mean that we need to separate between "actual maintenance mode" and "scheduled maintenance mode" (with the actual one being driven manually or through exit criteria (e.g. full maintenance task action chain has been completed successfully) and the scheduled on just indicating what _should_ be based on the schedule.

mcalmer replied:
To be able to plan, we need to know about maintenance windows upfront.
A 3rd party application which just set the maintenance mode "on" and "off"
is not helpful.

The emergency case is special. It can set the maintenance mode "now".
You cannot predict when an emergency happens, so there also no need to "plan"
actions for it.
When it happen, you set the system in maintenance mode and create new actions
wich should be exectuted "now".

This means when we check the schedule for maintenance windows we need to consider
the current state. If the system is currently in maintenance mode, we allow to schedule
actions "now" - maybe also for the next minute.

I will add a section about this case

hustodemon commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,99 @@
+## Implement a Maintenance Mode
+
+Check, how System Locking is implemented. Copy the mechanism for Maintenance Mode.
+Does it make sense to use System Locking as "Maintenance Mode"?
```
Publishing a comment about my research of system locking. There are 3 types of locking in Uyuni:

### Package locking
(in the web UI: `Software -> Packages -> Lock`)
- this allows locking some package versions on the system (probably uses zypper locking mechanism under the hood?)
- looks like we need something more general

### System locking (traditional systems only)

(in the Web UI, go to system details, click `lock system`)

- systems locked in such way will not allow running _some_ actions:
- the (black)list of forbidden actions is set in the DB: `select * from rhnActionType where unlocked_only = 'Y';`
- the blacklist is not customizable by the user
- all actions (even the blacklisted) can still be scheduled, when the system is locked
- the actions not in the blacklist are run normally
- as for the blacklisted actions: as soon as the system is unlocked, these scheduled actions are run (i.e. on next execution of the `rhn_check`, supposing the system is unlocked at the time of execution)
- implementation: system lock is an object in the database, there is a `1:1` relation with the system (or server) object
- **problems**
  - if we use it for the maintenance mode, we could collide with the existing locks (the admin can set the lock, independent of any maintenance window. Maybe we don't want to unlock the system that was locked explicitly by some administrator). This could be solved by documenting the behavior ("system locks can be overridden by the maintenance windows mechanism").

### Blackout mode (minions only)
- i.e. locking on "salt level"
- Implementation via a formula (see `system-lock-formula` RPM)
- when the formula is assigned to a system (group, org), the salt module calls will not be executed on the system(s) 
- difference to the system lock for trad. systems: when the formula is assigned, any actions run will **fail**
- allows whitelisting of allowed actions (`test.ping`, `grains.items`, `grains.item`)
- **problems**:
  - same as the previous case

### Q&A
- Q: we need to decide, how we should react when an action is executed outside of the maintenance window
  - A1: Postpone
  - A2: Fail
  - ~A3~: Don't allow executing that action at all. That would be very hard to achieve: actions can be scheduled for future. Maintenance window that forbids this action can be created **after** the action is scheduled.

- Q:  do we want to support both traditional clients and salt minions? if yes, we need to unify the fail/postpone semantics above
  - A: We focus primarily on minions. We could use the blackout mechanism for minions and go with the "fail" semantics above.

- Q: Can we actually use the blackout mechanism for minions?
  - pros:
    - already implemented mechanism
    - works on the salt modules level, which should be powerful and flexible enough
  - cons:
    - can collide with existing code that uses blacklisting (we would need a special handling for CaaSP nodes, which should always stay in blackout mode).
    - implementation via formula: it inherits the common formula integration drawbacks, which we might take into account (assignment on filesystems, not in the database: slower querying, no support for migration scripts)

mcalmer replied:
as we decided to not enforce the maintenance mode on the system itself in the first iteration I will mark this discussion as resolved.
This will come back when we discuss the 2nd iteration.

paususe commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,142 @@
+What to do?
+
+- Execute even after MU is over? (this is what may happen without changes)
+- New features are made for salt - just mention it in the docs and suggest to use osad, a short checking frequence or a very long Window?
```
This.

parlt91 commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,142 @@
+
+- In the systems details page we need to show the selected schedule with a view and a change option.
+- It should be possible to change the maintenance schedule for multiple systems using SSM
+- In the details of the Maintenance Schedule we could have a systems page where add/remove is possible.
```
As I understood it changing the schedule would change the schedule for all systems that follow that schedule.

parlt91 commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,142 @@
+- automatically choose next maintenance window to the given date?
+- set the datepicker to the next maintenance window?
+- change the datepicker into a simple combobox where we display the next maintenance windows
+
```
These datepickers make use of the `bootstrap-datepicker` component that does not allow excluding individual dates, so limiting these to just the ones inside of maintenance windows would be difficult.

mcalmer replied:
adapting the datepicker is tricky. Already suggest the combobox variant.

parlt91 commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,191 @@
+
+### Schedule an Action
+
+When a system is assigned to a schedule, only schedules withing a maintenance window should be allowed.
```
Should be: `...schedule, only actions within a ...`

mcalmer replied:
fixed

parlt91 commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,191 @@
+Systems in different schedules needs special handling:
+
+1. do not allow systems with different schedules in SSM
+2. divide systems by maintenance schedules and display a combo box with the names. Let the user choose for which set he want to schedule it. He need to repeat it for the rest.
```
Only concern here is that it might get messy to display the systems if many systems belong to either schedule

mcalmer replied:
I will add "next available" to the option.
It is a bit unspecified, but as we defined a resolve strategy for removing a window, it is a doable option.

parlt91 commented:
This looks already pretty good to me.
Double checking with my notes I found that we should probably mention something about the "Emergency mode" (allow scheduling actions for certain users under certain circumstances)

mbrookhuis commented:
can we define also a set of actions that should be executed during the maintenance and this should apply for all re-occurring appointments too. For actions think of:
- highstate
- remote command
- patch
- reboot
- highstate

And report the assigned admins (and when enabled) that the maintenance is finished or having issues.

joesusecom commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,191 @@
+- A flag in DB which needs to be changed?
+- We evaluate the calendar and calculate the state of the system?
+- mix: taskomatic job (every minute) which evaluate the calendar and switch the flag (and trigger any other action which is needed)
+
```
The part that's missing here is that systems that are using Maintenance Windows should probably be locked (like cluster nodes) while they're _not_ in maintenance mode. This will likely affect the implementation, as the locking needs to be applied and unapplied (so just doing a DB query `is_in_maintenance_mode` or so won't do the job.

mcalmer replied:
Well, we put exactly that to future iterations as this brings a lot of extra work.
When doing this in the first iteration we may not be able to deliver in time.

This is also the reason why it is in the open questions section.

On the other hand I am not 100% sure if this is really needed in the end.
Without it, it is a "SUSE Manager" feature and when you connect directly via salt nobody will hold you back on doing stupid things.
But maybe this is good enough for a lot of people. I think we will find out via feedback on this.

moio replied:
So do I get it correctly that this will be enforced for Actions only, and not at the Salt level?

hustodemon replied:
At least initially, yes. As written above, the "blackouting" (locking on the salt modules level) the minions **could** come up later (2nd iteration).

mcalmer replied:
Just found out that we might not be able to enforce this on minion level.
It would effect docker and kiwi image building. But these actions should happen when the system is **not** in maintenance mode.
If we really want to enforce it, we need a big refactoring and need to implement an own daemon to build images and exclude it from `salt`.

moio replied:
From a tech POV, I presently cannot imagine a unique solution to enforce this both at the Action level and the Salt level - I guess we will always somehow need two almost-separate code changes, as underlying mechanisms are completely different.

From a product POV, I also recognize this concept clicks much better with the Action model, and could be seen as added value on top of plain Salt.

On the other hand though, I see users having to or preferring to avoid Actions and work directly with Salt, at least in part, in many cases.

So my only recommendation would be: if possible, try to confirm from more-user-oriented colleagues that addressing Actions first is actually more valuable than having an (initial) Salt-only solution (I can easily imagine that being actually the case, just playing devil's advocate here).

joesusecom commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,191 @@
+- Execute even after MU is over? (this is what may happen without changes)
+- New features are made for salt - just mention it in the docs and suggest to use osad, a short checking frequency or a very long Window?
+- Special implementation?
+
```
My guidance for traditional clients is: Focus on Salt, but if there's no practical difference, I'm super-fine with getting the traditional stack support for free.

_If_ we end up being able traditional clients, my expectation (to be validated by someone in the field) would be that Uyuni strictly enforces the windows. So if a client checks in too late, bad luck and come back in the next window.

That said we'd most likely be able to just point to OSAD.

BTW, one thing we may want to check is whether admins are _currently_ using the run_check to implement their own "maintenance windows" (by only running the run_check when they actually want to do maintenance).

mcalmer replied:
This is similar to what I think. When customer uses osad, it should just work. We schedule for a time in the maintenance window and osad will trigger at that time.

Only when not using osad possible problems might happen.
But with documentation and long enough maintenance windows it should be doable.

parlt91 commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,191 @@
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+
```
What to do with **recurring actions**?
I see two possibilities:
1) Don't allow scheduling recurring events / disable existing events if there is a maintenance window defined.
2) Only execute the (recurring) action if the moment of execution is inside a defined maintenance window.

mcalmer replied:
thanks. added

mcalmer commented:
> can we define also a set of actions that should be executed during the maintenance and this should apply for all re-occurring appointments too. For actions think of:
> 
> * highstate
> * remote command
> * patch
> * reboot
> * highstate
> 
> And report the assigned admins (and when enabled) that the maintenance is finished or having issues.

You want actions which can be executed independent of the maintenance mode.
Well, this is tricky.

With the currently planned 1st iteration it would be possible to define "actions" which need to keep the maintenance mode into acount and "actions" which can be executed always at any time.

But when we start to enforce the maintenance mode on the client, we cannot differentiate anymore between "highstate" and "package list refresh" as both are a state.apply.

If this is wanted, we need to decide that maintenance mode is a "SUSE Manager" feature and pure salt can still overwrite everything.

mcalmer commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,191 @@
+[architecture]: #architecture
+
+A system is in "Maintenance Mode" or not.
+When a system is in "Maintenance Mode" no action can be executed.
```
totally right. I will change it

hustodemon replied:
I was also thinking like that subconsciously: having a system in _that_ mode means the system is "locked". Maybe it's because I'm used to the "lock" and "blackout" mode that this similar semantics.

I _really_ like the maintenance mode/window/schedule terminology and can't think of anything better with the "inverse" semantics (the best I could come up with, was "production" or "protected", but I don't like this, or "locked", which we already use), but maybe somebody else has a better idea. For me, the "maintenance" is still the best option.

mcalmer commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,191 @@
+
+A system is in "Maintenance Mode" or not.
+When a system is in "Maintenance Mode" no action can be executed.
+Setting the Maintenance Mode "on" and "off" requires Organization Admin permissions.
```
Sounds like an interesting idea.
I will add this to the open questions for now to have a discussion around it.

mcalmer commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,191 @@
+They can be defined as individual or as recurring events.
+
+To assign a Schedule to multiple systems, SSM should be used.
+
```
> If I'm not mistaken, this also means that we would easily be able to create API queries or
> UI views that show all systems that have a certain schedule assigned, right?

Yes, should be possible

I also added a "Second Iteration" section to the RFC and added there the activation key story.

mcalmer commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,191 @@
+A Maintenance Schedule is a set of Maintenance Windows. A Maintenance Schedule has a name/label.
+The data should be stored as ICalendar data (https://tools.ietf.org/html/rfc5545).
+In the first iteration importing the data created by an external tool is required.
+Uyuni will not provide a User interface for generating or manipulating Maintenance Windows.
```
I tried korganizer and office365 outlook. We should definitive test some of them.

mcalmer replied:
I add this to the question section.

mbologna commented:
An action chain is started in the maintenance window. Before all the actions in the action chains are executed, the maintenance window ends. What happens?

mcalmer commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,191 @@
+- Schedule Type (Single, Multi, ....)
+- Data (ICalendar content) with file upload option
+
+A "Create" button will create the new schedule. The ICalendar data are stored in the database.
```
I added this to "Second Iteration" to not forget about it.

mbologna commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,208 @@
+and scheduling actions inside of maintenance windows.
+
+Enforcing Maintenance Windows on lower level is part of the future iterations and will be defined later
+in updates of this RFC.
```
IIUC, this means that the implementation logic that actually blocks an action from being executed will be defined in future updates? I spoke with @hustodemon about Salt and traditional locking, and their conceptual differences. It might be a good occasion to unify the concept of "locking" in case we want to use it as a way to prevent action execution outside maintenance windows.

moio commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,390 @@
+Inside of the Schedule, multiple Maintenance Windows can be defined.
+They can be defined as individual or as recurring events.
+
+To assign a Schedule to multiple systems, SSM should be used.
```
SSM is a GUI-only concept. Can we expect an equivalent on the API side?

mcalmer replied:
will add XMLRPC function for this as well taking a list of systemids

moio commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,390 @@
+
+When a system is assigned to a maintenance schedule, actions which are flagged as `maintenance_mode_only` should be executed only within a maintenance window.
+In the User Interface we should show in these cases not the datepicker, but a simple Combo box with the Date and Time of the next 20 Maintenance Windows.
+All actions will be scheduled at the begin of the Maintenance Window.
```
If we constrain all `maintainance_mode_only` Actions to be scheduled at the beginning of a Maintenance Window, does it make sense to talk about Maintenance Windows at all? Is this constraint expected to be permanent, or to be lifted in a successive iteration?

If it is expected to be permanent, wouldn't it be easier to "rename" this whole concept to "Maintenance Start Times" and ignore lengths altogether?

(the only difference would be to prevent out-of-window scheduling via XMLRPC - arguably though if a user already uses XMLRPC he might have a scheduling already)

paususe replied:
> If we constrain all `maintainance_mode_only` Actions to be scheduled at the beginning of a Maintenance Window, does it make sense to talk about Maintenance Windows at all? Is this constraint expected to be permanent, or to be lifted in a successive iteration?

I don' think we should schedule everything at the beginning of the maintenance window, at least if there are multiple actions to be scheduled, i. e. do not overlap actions.

> If it is expected to be permanent, wouldn't it be easier to "rename" this whole concept to "Maintenance Start Times" and ignore lengths altogether?
> 
> (the only difference would be to prevent out-of-window scheduling via XMLRPC - arguably though if a user already uses XMLRPC he might have a scheduling already)

Maintenance windows are a well-established concept in the ITSM world, I don't think there's a need to redefine concepts or try to outsmart sysadmins.

In the real world, it works like this:

* "as my maintenance window is from 2 AM to 6 AM on Saturday and Sunday night, I will plan to start this maintenance procedure at 3 AM on Saturday and I [human] have estimated this will be finished at 4 AM, because I have done my testing in TEST and PRE [which in theory is identical to PRO] environments"
* Why 3 AM instead of 2 AM, which is the time the maintenance window starts? Because it turns out this week there is some other procedure at 2 AM, at the beginning of the maintenance window. It does not even need to be for this server, but given that a human operator is always supervising or executing changes, a procedure on server A at 2 AM may prevent another procedure on server B at 2 AM due to lack of humans
* If for some reason (human or machine) the procedure does not finish on time, there are two possibilities:
  * Rollback is possible and/or desirable: a final "go/no-go" decision is made at e. g. 5 AM (whatever is enough to cancel the procedure and perform the rollback)
  * Rollback is not possible and/or undesirable: let the procedure finish at e. g. 7 AM, and register an out-of-window event (incident) in the ITSM tool (e. g. ServiceNow or BMC), calculate SLIs/SLAs, etc

moio replied:
> I don' think we should schedule everything at the beginning of the maintenance window, at least if there are multiple actions to be scheduled, i. e. do not overlap actions to be executed on the same client.

Note: both the traditional and the Salt stack queue execution of Actions if they were to be concurrently on the same client.

mbrookhuis replied:
Subscribe to (software) channels should always be possible.

mcalmer replied:
It seems there is a preference to make it at least possible to schedule an action later in the maintenance window. I think I have a good idea how to make it easy possible.

We would implement a combobox with the next 20 maintenance windows and additionally an entry field with "Offset in Minutes" which has the default `0`. Using the offset, you can schedule an action 60 Minutes after the maintenance window starts. I also think remember an offset is easier than an exact time.

I will add this to the RFC

mcalmer replied:
> Subscribe to (software) channels should always be possible.

Will add this to the unresolved questions and we decide in the afternoon.

moio replied:
Just a reminder: changing software channel subscriptions for minions might imply product package installation!

mcalmer replied:
But only when a highstate is called. (Traditional clients are different, but they are somehow excluded when thinking about this feature)

moio replied:
Right, I forgot about that. Should not be a worry then.

moio commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,390 @@
+Adding a Maintenance Schedule via XMLRPC API must be possible as well.
+
+
+### Updating a Maintenance Schedule
```
```suggestion
### Updating/Deleting a Maintenance Schedule
```

mcalmer replied:
Removing a whole schedule is easy, as it does not invalidate any scheduled action. No schedule means permanent maintenance mode. You can execute actions all the time.

But I will add this info to this section.

moio commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,390 @@
+- In the details of the Maintenance Schedule we could have a systems page where add/remove is possible.
+
+If a Maintenance Schedule is assigned and which one, should be visible in the systems details page.
+It should be also part of the `system.getDetails()` XMLRPC API function.
```
Or possibly a separate endpoint, as `getDetails` contains already quite a lot of information.

mcalmer replied:
Changed that part and name also "setting" the schedule via API and name also SSM.
I have removed the explicit naming of this function. We will decide at implementation time what is the best thing to do.

moio commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,390 @@
+
+The current status of the system should be reflected in the User Interface and API.
+
+If a system is in Maintenance Mode, the User Interface could show a special icon [ || ] (pause symbol of an audio player).
```
Not sure if such an icon is a good idea, it could be ambiguous with virtualization pausing which some hypervisors support. I agree with the general idea to have it reflected in the UI somehow, just maybe we will want a different icon or just a label.

mcalmer replied:
Noted. Will discuss in the afternoon.

moio commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,390 @@
+
+When a system is assigned to a maintenance schedule, every finished action should check, if the system is still in maintenance mode.
+If this is not the case it means, the action was started inside of the maintenance window, but was not able to finish in time.
+This Maintenance Window overflow should be show with a notification.
```
Ideally we will not have one notification per system, or those could easily overflow the UI (imagine a big batch of systems all more-or-less simultaneously missing their window).

mcalmer replied:
added to the RFC.

moio commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,390 @@
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- performance: what happens when a lot of actions are scheduled at the beginning of a maintenance window? Will this create a performance problem? Is salt batching ready to handle such situations properly?
```
We should be fine as long as we create one Action per type and timestamp with multiple clients associated. Specifically, we should try to avoid as much as possible the situation in which there is one Action per client.

Batching will limit the number of clients executing an Action (only considering that one Action).

My impression is that this proposal would not make the situation worse than it is now - of course users will still have to pay a bit of attention not to create too many separate Actions or batching won't help them. But by and large, they should be able to create Actions targeting many clients just like before - just "taking one pre-defined start datetime from a list" instead of a specifying one "freely".

We will have to pay attention when, for example, rescheduling Actions due to updates to their Maintenance Schedule - clients will have to stay grouped together as much as possible.

moio commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,390 @@
+- Nearest: Use the nearest maintenance window
+- Previous: Use the maintenance window before the scheduled date
+- Next: Use the maintenance window after the scheduled date
+- Cancel: Cancel the Action
```
Would it be possible to only implement a subset of these, for example only Cancel, for a first version?

I might be wrong, but in my eyes this functionality is quite an important complexifier.

mcalmer replied:
I think it is possible to implement only some of them. But I am not sure if this is the issue.
I see the general check for all scheduled actions as the bigger problem.

We will see how far we can get.

hustodemon commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,409 @@
+
+## Notifications
+
+When a system is assigned to a maintenance schedule, every finished action should check, if the system is still in maintenance mode.
```
Moreover, we need to check, if the system is still in the same maintenance **window**. What could happen, is that a long running action could overflow to a next maintenance window and everything would look ok.

mcalmer replied:
From my point of view not a hard requirement. At least not for the first iteration.
I think a typical customer define for a system only 1 window per week.
If that overflow to the next week without anybody notice it, they have a different problem.

hustodemon commented on accepted/00000-maintenance-windows.md:
```diff
@@ -0,0 +1,409 @@
+
+## Second Iteration
+
+### Implement a Maintenance Mode
```
"Maintenance mode" has been mentioned multiple times before, and now there is "Implement a Maintenance Mode" section. I think it'd be better to explicitly mention, that this is the maintenance mode "override" (or "flag").

```suggestion
### Implement a Maintenance Mode override
```

mcalmer commented:
> Some final comments:
> 
> * Being able to toggle (switch on/off) the maintenance mode (without having to assign a new schedule) seems to be a very important feature to me, especially when it comes to integrating with tools like ServiceNow or check_mk. Fine with labeling it phase two, but let's address it soon after, before putting more effort into other "refinements".

I will create an issue for it which can be scheduled by Pau

> * The "resolve strategy" for handling scheduled actions that wouldn't fit into a window after the schedule has changed seems overly ambitious/complicated to me. No text changes required, but let's make sure we don't get held up by this. In most cases, admins will know what they're doing, and having to reschedule tasks manually after the maintenance window has changed sounds like a perfectly valid approach.

We will try to keep the time short here. But remember that you cannot "re-schedule" an action from UI or API. You can only cancel it and re-create it. So the idea was for MVP to just offer "Cancel" but provide a list of actions which will be canceled. With this list the admin could hopefully re-create them.

If time permits, we may implement also another strategy. Maybe also as Maintenance Update.

> * One thing that doesn't become very clear in the RfC is that adding/removing a schedule basically also means that a system changes from being unprotected from admins to being protected during the non-scheduled times. We should make that clear both in naming the APIs and in the UI.

Added to doc issue

> * Maintenance mode vs. locking: It could be very confusing to explain what happens to a system that is both locked and uses maintenance windows. I'm fine per se with keeping those two concepts separate, but we really need to document well (both in the status UI and the actual documentation) what one can do to a system at any given point in time and who can change the mode/lock how.

Added to doc issue

> * Can we please put the list of actions that are allowed/blocked into the documentation? This is great stuff, but shouldn't require a database inspection tool to verify. :-)

Added to doc issue

> * the virt.* actions and kickstart_guest.initiate functions need to be excluded from the maintenance-only mode: Running VMs is production use for a VM host!

Done

> * We'll need to plan for a follow-up on handling VMs. For example, as long as we don't have automated live migration of VMs, a VM will automatically be affected by maintenance mode when the host is rebooted or becomes unstable because of ongoing package updates! So, from an SLE perspective, _all_ VMs on a host that's under maintenance may be down! On the other hand, the VMs can't just inherit the schedule because
>   
>   * when the host is down, you _can't_ actually update them
>   * you may want to have maintenance windows for them at times when the host is definitely up!

Added to 2nd iteration epic as topic (to not forget it)

> * For scheduling actions, we really need to have another RfC related to this that handles storing re-usable action chains. Having to schedule tasks one by one with enough "safety margin" is just painful.

We have already an issue for it https://github.com/SUSE/spacewalk/issues/3439 . It was created when we made the first step for action chains. I will just add it to the planning board.

> * Idea for later: We could display all scheduled actions and windows in a web calendar at some point!

Added to 2nd Iteration epic

mcalmer commented:
> Add-on comments on the 2nd iteration chapter:
> 
> **Maintenance Mode**
> The "Maintenance Mode override" only covers one of the two use cases: Emergencies. The other one is when a third party tool (e.g. the monitoring tool, service management application, or CMDB) actually controls the maintenance modes. In that case, instead of having to export calendars (which those tools may not even support), we'd need an action API anyway that allows one to toggle the maintenance mode for one or more systems.
> 
Well, but without calendar you cannot schedule anything. This is a problem. Such a Tool could enable the maintenance mode, but nothing is prepared at that point in time. So this makes sense only when the admin sitting in front of SUSE Manager and execute the actions "now".

Other option would be to change the recuring action feature. You prepare a recurring action which run automatically when the maintenance window is turned "on".
Here we need to enhance the RFC a bit more.

I just added this to the RFC
> 
> **Salt**
> Whether or not to enforce the maintenance mode on the Salt side is a design question that goes beyond this particular topic. We haven't implemented ACLs for Salt either. So what we need to discuss is whether it's generally fine even in the future that only actions that go _through_ SUSE Manager (UI or API) are protected by our policies (ACLs, maintenance windows, locking settings, cluster awareness, ...) or whether we also need to implement them for _native_ Salt (salt CLI, Salt API) at some point.

Aggreed, needs further discussions and we need to bring more people into it.

> **Native editing**
> While using the iCalendar approach is a great design idea, we may also want to integrate an actual schedule editor at some point (which could be just a generic calendar plugin that translates the UI input into an iCalendar file).

Added to RFC 2nd It.

> **Visualizing the calendars**
> We could have added a sentence or two about the plan to visualize calendars.

Added to RFC 2nd It.
